File: 2018/day04.php
<?php

class Log
{
    public $when;
    public $guardId;
    public $beginShift = false;
    public $fallsAsleep = false;
    public $wakesUp = false;

    public function __construct(DateTimeImmutable $when, bool $beginShift, bool $fallsAsleep, bool $wakesUp, ?int $guardId = null)
    {
        $this->when = $when;
        $this->beginShift = $beginShift;
        $this->fallsAsleep = $fallsAsleep;
        $this->wakesUp = $wakesUp;
        $this->guardId = $guardId;
    }
}

$fh = fopen(__DIR__.'/inputs/04.txt', 'r');
$logs = [];

// parse log entries
while (!feof($fh)) {
    $line = trim(fgets($fh));
    if (empty($line)) {
        continue;
    }

    $matches = [];
    $matched = preg_match('/^\\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2})\\] (.*)$/', $line, $matches);

    if ($matched !== 1) {
        continue;
    }

    $when = DateTimeImmutable::createFromFormat('Y-m-d H:i', $matches[1]);
    $beginShift = false;
    $fallsAsleep = false;
    $wakesUp = false;
    $guardId = null;

    if ($matches[2] === 'wakes up') {
        $wakesUp = true;
    } else if ($matches[2] === 'falls asleep') {
        $fallsAsleep = true;
    } else {
        $beginShift = true;
        $guardId = intval(substr(explode(' ', $matches[2])[1], 1));
    }

    $logs[] = new Log($when, $beginShift, $fallsAsleep, $wakesUp, $guardId);
}

// sort by when the log entry was made
usort($logs, function(Log $a, Log $b) {
    return $a->when <=> $b->when;
});

// find the sleepiest guards by minutes slept
$sleepiestGuards = [];
$currentGuardId = $logs[0]->guardId;
$fellAsleep = null;

foreach ($logs as $l) {
    if ($l->guardId === null) {
        $l->guardId = $currentGuardId;
    }

    if ($l->beginShift) {
        $currentGuardId = $l->guardId;
    }

    if ($l->fallsAsleep) {
        $fellAsleep = $l->when;
    }

    if ($l->wakesUp) {
        $timeAsleep = intval(($l->when->diff($fellAsleep))->format('%i')) - 1;

        if (!array_key_exists($l->guardId, $sleepiestGuards)) {
            $sleepiestGuards[$l->guardId] = 0;
        }

        $sleepiestGuards[$l->guardId] += $timeAsleep;
    }
}

asort($sleepiestGuards, SORT_NUMERIC);
$sleepiestGuards = array_reverse($sleepiestGuards, true);
$sleepiestGuardId = key($sleepiestGuards); // array cursor is rewound

// get all of the log entries pertaining to the sleepiest guard
$sleepyLogs = array_filter($logs, function(Log $l) use ($sleepiestGuardId) {
    return $l->guardId === $sleepiestGuardId;
});

// find the sleepiest minute
$sleepyMinutes = array_fill(0, 59, 0);
$fellAsleep = null;

foreach ($sleepyLogs as $l) {
    if ($l->fallsAsleep) {
        $fellAsleep = $l->when;
    }

    if ($l->wakesUp) {
        $minStart = intval($fellAsleep->format('i'));
        $minEnd = intval($l->when->format('i'));

        for ($i = $minStart; $i < $minEnd; $i++) {
            $sleepyMinutes[$i]++;
        }
    }
}

asort($sleepyMinutes, SORT_NUMERIC);
$sleepyMinutes = array_reverse($sleepyMinutes, true);

var_dump(key($sleepyMinutes) * $sleepiestGuardId);

File: 2018/day04.py
import re
from datetime import datetime

class Log:
    def __init__(self, datetime, guard_id, start, awake, asleep):
        self.datetime = datetime
        self.guard_id = guard_id
        self.start = start
        self.awake = awake
        self.asleep = asleep

class Guard:
    def __init__(self, id, total, minutes):
        self.id = id
        self.minutes = minutes

def parse(line):
    m = re.search('^\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2})\] (wakes up|falls asleep|Guard #(\d+) begins shift)$', line)
    dt = datetime.fromisoformat(m.group(1))
    guard_id = None
    start = False
    awake = (m.group(2) == 'wakes up')
    asleep = (m.group(2) == 'falls asleep')

    if not awake and not asleep:
        start = True
        #awake = True
        guard_id = int(m.group(3))

    return Log(dt, guard_id, start, awake, asleep)

with open('inputs/04.txt', 'r') as f:
    lines = f.read().splitlines()
    logs = [parse(line) for line in lines]
    logs.sort(key=lambda v: v.datetime)

    current_guard = None
    asleep = None
    guards = {}

    for log in logs:
        if log.start:
            if log.guard_id not in guards:
                current_guard = Guard(log.guard_id, 0, [0 for i in range(0, 60)])
                guards[log.guard_id] = current_guard
            else:
                current_guard = guards[log.guard_id]

        if log.asleep:
            asleep = log.datetime

        if log.awake:
            for minute in range(asleep.minute, log.datetime.minute):
                current_guard.minutes[minute] += 1

            asleep = None

    sorted_guards = [ guard for guard in guards.values() ]
    sorted_guards.sort(key=lambda v: sum(v.minutes), reverse=True)

    sleepy_guard = sorted_guards[0]

    part_1 = (sleepy_guard.id, sleepy_guard.minutes.index(max(sleepy_guard.minutes)))

    max_sleep = 0
    part_2 = ()

    for guard in guards.values():
        most_sleep = max(guard.minutes)
        if most_sleep > max_sleep:
            part_2 = (guard.id, guard.minutes.index(most_sleep))
            max_sleep = most_sleep

    print(part_1[0] * part_1[1], part_2[0] * part_2[1])
